Minio cluster becomes unresponsively if a minority of disks are down

I have a custom Docker image where I bundle Minio and MC [docker.tar.gz](https://github.com/minio/minio/files/1474050/docker.tar.gz). Beside this, this Docker image contains a startupHeal script that will do the healing when the nodes joins the cluster, if it was down due to different reasons (since I will use Kubernetes for orchestration, the nodes can be re-started automatically). This scripts waits for a 30 seconds before starting looking if the cluster is up (all the nodes don't have errors) and then proceed with a basic heal operation. 

I start a Minio cluster using [docker-compose.txt](https://github.com/minio/minio/files/1474051/docker-compose.txt) and upload some files with [MinioClientUploader.txt](https://github.com/minio/minio/files/1474054/MinioClientUploader.txt). Everything works as expected, but when I kill a node (for example minio3), the cluster becomes unresponsively and cannot upload files using the Java client. More, the UI and `mc ls` sometimes display `disk error` and sometimes responds very slow. This ticket is very similar with https://github.com/minio/minio/issues/5189, but the difference is that we are using a custom image (with Minio, MC and a startupHeal script bundled) and the result of uploads during un-complete cluster is an error sometimes. 

## Your Environment
* Version used (`minio version`): Minio - latest, MC - latest
* Operating System and version (`uname -a`): 42~16.04.1-Ubuntu

HI @krisis, I performed more tests and come back with a piece of code that throws the exception: 
[MinioClientUploaderV2.txt](https://github.com/minio/minio/files/1501525/MinioClientUploaderV2.txt). This is a Java code, so please change the extension from `.txt` to `.java`. 
So, using the initial docker image (the one from the initial post, with Minio and MC), While running the attached Java code which will upload a file multiple times and does also a `listObjects`, I'm killing one of the three nodes from the cluster (from the initial docker-compose file - for example minio3). As a result, the file is still uploaded, but I get the following stacktrace while getting the bucket objects:
`java.net.SocketTimeoutException: timeout
	at okio.Okio$4.newTimeoutException(Okio.java:227)
	at okio.AsyncTimeout.exit(AsyncTimeout.java:284)
	at okio.AsyncTimeout$2.read(AsyncTimeout.java:240)
	at okio.RealBufferedSource.indexOf(RealBufferedSource.java:344)
	at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:216)
	at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:210)
	at okhttp3.internal.http1.Http1Codec.readResponseHeaders(Http1Codec.java:189)
	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:67)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:120)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:179)
	at okhttp3.RealCall.execute(RealCall.java:63)
	at io.minio.MinioClient.execute(MinioClient.java:998)
	at io.minio.MinioClient.executeGet(MinioClient.java:1192)
	at io.minio.MinioClient.listObjectsV2(MinioClient.java:2609)
	at io.minio.MinioClient.access$100(MinioClient.java:158)
	at io.minio.MinioClient$5$1.populate(MinioClient.java:2483)
	at io.minio.MinioClient$5$1.hasNext(MinioClient.java:2506)
	at com.example.MinioClientUploader.main(MinioClientUploader.java:35)
Caused by: java.net.SocketException: Socket closed
	at java.net.SocketInputStream.read(SocketInputStream.java:204)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at okio.Okio$2.read(Okio.java:138)
	at okio.AsyncTimeout$2.read(AsyncTimeout.java:236)
	... 26 more`

Also the upload performs quite slow, so this might be the reason why I get that timeout exception. 
If I try to perform an "mc ls local/photos` I got `mc: <ERROR> Unable to list folder. We encountered an internal error, please try again.`

Please let me know if you are able to reproduce the issue with the provided Java code.

Thanks for the help and hopefully we can pinpoint the issue,
George

--- a/cmd/admin-rpc-server.go
+++ b/cmd/admin-rpc-server.go
@@ -104,8 +104,12 @@ func (s *adminCmd) ReInitDisks(args *AuthRPCArgs, reply *AuthRPCReply) error {
 		return err
 	}
 
+	// Wrap into retrying disks
+	retryingDisks := initRetryableStorageDisks(bootstrapDisks,
+		time.Millisecond, time.Millisecond*5, globalStorageHealthCheckInterval, globalStorageRetryThreshold)
+
 	// Initialize new object layer with newly formatted disks.
-	newObjectAPI, err := newXLObjects(bootstrapDisks)
+	newObjectAPI, err := newXLObjects(retryingDisks)
 	if err != nil {
 		return err
 	}

